src/couch_views/src/couch_views_batch.erl

```diff
@@ -0,0 +1,192 @@
+% Licensed under the Apache License, Version 2.0 (the "License"); you may not
+% use this file except in compliance with the License. You may obtain a copy of
+% the License at
+%
+% http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+-module(couch_views_batch).
+
+
+-export([
+    start/0,
+    success/2,
+    failure/0
+]).
+
+-export([
+    start/1,
+    success/3,
+    failure/1
+]).
+
+
+-callback start(State::term()) -> {NewState::term(), BatchSize::pos_integer()}.
```

Comment:
Should we define and use `-opaque state() :: term().`?

Reply:
I'm not sure what that buys us or how it'd interact with other modules implementing the behavior?

Reply:
With you latest rewrite and separation into two modules it is not required to use `opaque`.